### ISSUE TITLE
[apex] New Rule: Avoid Stateful Database Results

### ISSUE THREAD

**Issue by mitchspano** [pmd/pmd#5305]
**Proposed Rule Name:** [apex] Avoid Stateful Database Results

**Proposed Category:** Error Prone

**Description:** Instance variables of type `Database.[x]Result` can cause serialization issues between successive batch iterations when used in an implementation of the `Database.Stateful` interface. Please avoid using such variables unless they are transient.

This helps to avoid [this known issue](https://issues.salesforce.com/issue/a028c00000qPwlqAAC/stateful-batch-job-that-stores-databasesaveresult-failed-after-validation-errors-throws-error-during-deserialization)

**Code Sample:**

```java
public class Example implements Database.Batchable, Database.Stateful {
  Database.SaveResult[] results; // this can cause failures 

  public Database.Querylocator start(Database.BatchableContext context){
    return Database.getQueryLocator('SELECT Id FROM Account LIMIT 10');
  }

  public void execute(Database.BatchableContext context, List<SObject> scope){
    results = Database.update(scope, false);
  }

  public void finish(database.BatchableContext context){
  }
}
```

**Possible Properties:**

*   Should this rule be customizable via properties? No, this is a straightforward rule.

**Comment by pkozuchowski** [comment]
Downvoted because this sounds more like a Salesforce bug, than code smell or design issue, and not a common one too. Considering how many records batch can process, storing all save results may lead to heap space limit, so maybe it would be better to have a "pay attention" to heap space rule in Statetful batches.

**Comment by adangel** [comment]
@stephen-carter-at-sf @rsoesemann 
Do you think, this rule is useful/makes sense? There is already a PR that would implement this rule... (see #5425 )

**Comment by FishOfPrey** [comment]
Note that this is a very specific issue around how batch jobs serialized/deserialize Database.SaveResult records and the like related to validation rules. It is so specific that in 2018 it didn't warrant a platform fix. There are viable workarounds related to either using JSON serialization or another structure to hold the required data.

**Comment by johnbelosf** [comment]
@adangel as per @FishOfPrey 's comment above, we don't think this needs to be prioritized

**Comment by mitchspano** [comment]
I would argue that @FishOfPrey's response makes this _more_ important to implement as a static analysis rule.

While this issue might not occur frequently, its impact is severe. Because the problem is subtle and related to Salesforce's internal serialization of `Database.[x]Result` in batch Apex, it's extremely difficult and time-consuming to diagnose and resolve.

### Intuitive but Flawed Approach
It is very reasonable and intuitive for a developer to imagine using a stateful `Database.[x]Result` variable to track unsuccessful DML operations throughout a batch's execution, then report or handle them during the `finish` method. However, this can actually expose developers to a frustrating and difficult-to-debug situation.

### Intermittent and Vague Error
The error doesn't show up deterministically on every run; in my experience it was something like 10-15% of batch runs would encounter the error. When the error does arise, it is not reported immediately to the user. Sometimes it takes multiple hours after the batch was initiated for the batch executor to receive an email. The email they receive is cryptic at best; just an "Internal Salesforce.com Error" message and nothing else to help them identify the issue.

### Lengthy and Frustrating Debugging
Developers often next face a frustrating back-and-forth with Salesforce support, who may or may not be able to pinpoint the root cause. This entire process can take days.

### Lack of Developer Awareness and Clear Guidance
Most developers are unaware of the underlying serialization issue with `Database.[x]Result` in stateful batch Apex. Even the official documentation on the known issue page only provides a vague workaround: "Use a more generic type like map." 

### <ins>This is the very essence of "error-prone" code that we hope to catch with static code analysis.</ins>

The intermittent nature and vague error messages make it a debugging nightmare. The nature of this problem is very specific, which makes it easy to statically identify and ensures that minimal false positives will be reported. This proposed rule provides concrete guidance by flagging the problematic `Database.[x]Result` usage and offers a clear, documented solution, such as the provided `StatefulResult` example. 

Given that Salesforce has been aware of this issue since 2018 without a platform fix, it's even more important to provide developers with tools to avoid this landmine.

**Comment by adangel** [comment]
Sorry for the late reply.

> It is so specific that in 2018 it didn't warrant a platform fix.

This is actually an argument for having this rule - since this is a real problem that is not fixed. While the problem might be very specific, it still exists.

So I'm planning to merge this for 7.11.0.

@mitchspano Thank you for the detailed explanations!
